pkgs/development/compilers/gambit/build.nix

```diff
@@ -38,6 +38,10 @@ stdenv.mkDerivation rec {
       #--enable-inline-jumps
       #--enable-char-size=1" ; default is 4
     )
+    export CC=${gcc}/bin/gcc CXX=${gcc}/bin/g++ \
```

Comment:
MacOS's stdenv uses clang by default.  Depending on `gcc` means that using gambit might need to build all of gcc.

`$CC` should already be set, although without a full path, meaning you should be able to say `export CC="$(command -v $CC)" CXX="$(command -v $CXX)" ...` and so forth, and not depend on gcc.  I did something like this for plan9port
[here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/system/plan9port/builder.sh#L18).

This way, the reference to gcc can be removed.

Reply:
I don't remember how much slower clang-generated code it, but it's a significant factor. 200MB of space is cheap these days, speed is not. If you want a slow Scheme, there is plenty of competition to Gambit.

Can you try the two recipes and compare the generated code on some benchmarks?

And even if we pick clang, we probably want the path hardcoded with some `CC=` statement, instead of randomly inherited from the `$PATH`.

Reply:
Note that the `CC=` statement, etc., is important, because IIUC, the values are persisted in the installation as default values.

Reply:
(Thus gambit knows where to find a C toolchain that is compatible with what it used to compile its own code.)

Reply:
I think I have seen comparisons of latest-at-the-time releases where Clang was generating noticeably faster code. For some input code bases and target CPUs… I think these comparisons coexisted with the opposite comparisons for the same compiler versions.

I would not be surprised if Clang is 10% faster (for the current versions in Nixpkgs). Or 10% slower.

Reply:
Quoting a message by Marc Feeley on the gambit-list from 2018-05-05 (not the latest, but AFAIK, things haven't changed dramatically since then):
```
GCC is still the best choice when compiling Gambit.

When configured with the usual

  ./configure --enable-single-host

the build time of a “make -j8” on my 4 core laptop is over 10x slower when using CLANG.  As for speed of the generated code, CLANG produces code that is about 3x slower than GCC.

Here are the raw results, in seconds:

                    make -j8     test4
 GCC 8.1.0           39.566      1.167
 CLANG 802.0.38     426.371      4.229

When configured with

  ./configure --enable-single-host --enable-c-opt

which uses -O2 rather than the default -O1, the results are also favourable for GCC.  However the build times are closer and, interestingly, both compilers generate slightly slower code with -O2:

                    make -j8     test4
 GCC 8.1.0          193.560      1.218
 CLANG 802.0.38     229.945      4.410
```

Reply:
So not only is clang significantly slower, it generates code that is 3x slower!

Reply:
I don't remember the numbers that were quoted at the recent Gambit@30 conference, but I remember that they were also skewed significantly in favor of GCC.

I'm sure that Gambit and/or Clang could be fixed to make the combination faster, but that's work that no one has done yet.

Reply:
Wait, GCC -O2 is also slower than GCC -O1? In addition to Clang taking longer than GCC to compile which is … not the expected situation. Maybe GCC-efficient support for unrestricted reenterable continuations leads to unique idioms.

Reply:
Yes, now I need to compile with -O1, -Os, -O2 and see what works better. Until I do my own benchmark, I should probably follow Marc's advice and leave it at -O1.

Reply:
Maybe it is even worth a comment that the Gambit codebase manages to violate the basic assumptions about the compiler performance and compiled-code performance…

Reply:
I will, in my next PR (or this one, if still not merged), after I confirm the benchmarks...
It takes over 1h to run the benchmarks at https://github.com/ecraven/r7rs-benchmarks, plus maybe half an hour to recompile gambit, so say 2h per configuration, 4h for the next two configurations (-O1 and -Os)...

Reply:
In my benchmarks, `-Os` clearly wins over both `-O1` and `-O2` most of the time (as far as the generated code goes), so that's what I'll use.